File: code/packages/swift/b-tree/Sources/BTree/BTree.swift
/// BTree.swift — B-Tree (DT11) implementation in Swift
/// ======================================================
///
/// # What is a B-Tree?
///
/// A B-Tree is a self-balancing search tree designed to work efficiently on
/// disk-based storage systems.  Unlike a binary search tree (which has at most
/// 2 children per node), a B-Tree node can hold many keys and many children.
/// This "wide and shallow" shape keeps the height tiny — even a tree with a
/// billion entries typically has a height of just 3 or 4.
///
/// Invented by Rudolf Bayer and Edward McCreight at Boeing Research Labs in 1970,
/// B-Trees are the foundation of almost every database and filesystem you will
/// ever use: SQLite, PostgreSQL, MySQL InnoDB, NTFS, HFS+, ext4 all rely on
/// B-Tree variants under the hood.
///
/// # The minimum-degree parameter t
///
/// Every B-Tree is parameterised by an integer t ≥ 2 called the **minimum degree**.
/// It governs how fat each node is:
///
///   • A non-root node must hold at least t-1 keys  (so it is "at least half full")
///   • Every node holds at most 2t-1 keys            (so it never overflows)
///   • A non-leaf node with k keys has exactly k+1 children
///
/// When t = 2 we get a **2-3-4 tree**: nodes hold 1, 2, or 3 keys.
///
/// # Invariants — the rules the tree must NEVER break
///
/// 1. All leaf nodes sit at the same depth  (the tree is perfectly balanced).
/// 2. Every node's keys are in strictly ascending order.
/// 3. Every non-root node has at least t-1 keys.
/// 4. Every node has at most 2t-1 keys.
/// 5. For an internal node with keys [k0, k1, …, km], the i-th child subtree
///    contains only keys x such that k_{i-1} < x < k_i.
///
/// # Insertion — proactive top-down splitting
///
/// We use the "single-pass" algorithm: as we walk down the tree searching for
/// the insertion point we pre-emptively split any node that is already full
/// (has 2t-1 keys).  This means we never have to backtrack upwards.
///
/// # Deletion — CLRS top-down approach
///
/// We use the CLRS (Introduction to Algorithms) top-down deletion algorithm.
/// Before recursing into any child C we ensure C has at least t keys (one more
/// than the minimum), so that if we need to delete a key from C, C can afford
/// to lose one key and still satisfy the invariant.
///
/// The key insight: we call `_deleteFromNode` only on nodes that already have
/// ≥ t keys (guaranteed by the caller), except for the root (which can have
/// as few as 1 key).
///
/// ASCII diagram — splitting a full node when t=2:
///
///        P: [10, 30]                      P: [10, 20, 30]
///           /   |   \          →              /  |   |   \
///          A  [15,20,25]  B               A  [15] [25]  B
///                                  (median 20 rises into P)

// ─────────────────────────────────────────────────────────────────────────────
// BTreeNode — the building block
// ─────────────────────────────────────────────────────────────────────────────

/// A single node in a B-Tree.
final class BTreeNode<K: Comparable, V> {
    var keys: [K]
    var values: [V]
    var children: [BTreeNode<K, V>]
    var isLeaf: Bool

    init(isLeaf: Bool) {
        self.keys = []
        self.values = []
        self.children = []
        self.isLeaf = isLeaf
    }

    /// Find the first index i where keys[i] >= key.
    func findKeyIndex(_ key: K) -> Int {
        var i = 0
        while i < keys.count && keys[i] < key { i += 1 }
        return i
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// BTree — the public API
// ─────────────────────────────────────────────────────────────────────────────

/// A generic B-Tree that maps keys of type K to values of type V.
///
/// Usage example:
/// ```swift
/// var tree = BTree<Int, String>()
/// tree.insert(10, "ten")
/// tree.insert(20, "twenty")
/// let v = tree.search(10)   // → "ten"
/// let r = tree.rangeQuery(from: 5, to: 15)  // → [(10, "ten")]
/// ```
public final class BTree<K: Comparable, V> {
    private let t: Int
    private var root: BTreeNode<K, V>
    private var _count: Int = 0

    /// Create an empty B-Tree with the given minimum degree.
    /// - Parameter t: Minimum degree (default 2).  Must be ≥ 2.
    public init(t: Int = 2) {
        precondition(t >= 2, "Minimum degree t must be at least 2")
        self.t = t
        self.root = BTreeNode(isLeaf: true)
    }

    // ── Public properties ────────────────────────────────────────────────────

    public var count: Int { _count }
    public var height: Int { _height(node: root) }

    // ── Search ────────────────────────────────────────────────────────────────

    /// Return the value for `key`, or `nil` if the key is not present.
    public func search(_ key: K) -> V? {
        return _search(node: root, key: key)
    }

    /// Return true if `key` is present in the tree.
    public func contains(_ key: K) -> Bool {
        return search(key) != nil
    }

    // ── Insert ────────────────────────────────────────────────────────────────

    /// Insert or update the mapping key → value  (upsert semantics).
    public func insert(_ key: K, _ value: V) {
        if root.keys.count == 2 * t - 1 {
            let newRoot = BTreeNode<K, V>(isLeaf: false)
            newRoot.children.append(root)
            _splitChild(parent: newRoot, childIndex: 0)
            root = newRoot
        }
        let didInsert = _insertNonFull(node: root, key: key, value: value)
        if didInsert { _count += 1 }
    }

    // ── Delete ────────────────────────────────────────────────────────────────

    /// Remove the key from the tree.
    /// - Returns: `true` if the key was found and removed.
    @discardableResult
    public func delete(_ key: K) -> Bool {
        guard _count > 0 else { return false }
        let found = _delete(node: root, key: key)
        if found {
            _count -= 1
            if root.keys.isEmpty && !root.isLeaf {
                root = root.children[0]
            }
        }
        return found
    }

    // ── Range / traversal ────────────────────────────────────────────────────

    public func minKey() -> K? {
        guard !root.keys.isEmpty else { return nil }
        var node = root
        while !node.isLeaf { node = node.children[0] }
        return node.keys[0]
    }

    public func maxKey() -> K? {
        guard !root.keys.isEmpty else { return nil }
        var node = root
        while !node.isLeaf { node = node.children[node.children.count - 1] }
        return node.keys[node.keys.count - 1]
    }

    public func rangeQuery(from low: K, to high: K) -> [(K, V)] {
        var result: [(K, V)] = []
        _rangeQuery(node: root, low: low, high: high, result: &result)
        return result
    }

    public func inorder() -> [(K, V)] {
        var result: [(K, V)] = []
        _inorder(node: root, result: &result)
        return result
    }

    // ── Validation ────────────────────────────────────────────────────────────

    public func isValid() -> Bool {
        return _isValid(node: root, minKey: nil, maxKey: nil,
                        expectedDepth: height, currentDepth: 0, isRoot: true)
    }

    // =========================================================================
    // MARK: — Private implementation
    // =========================================================================

    // ── Search helper ─────────────────────────────────────────────────────────

    private func _search(node: BTreeNode<K, V>, key: K) -> V? {
        let i = node.findKeyIndex(key)
        if i < node.keys.count && node.keys[i] == key { return node.values[i] }
        if node.isLeaf { return nil }
        return _search(node: node.children[i], key: key)
    }

    // ── Split helper ──────────────────────────────────────────────────────────

    /// Split the i-th child of `parent` (which must be full — 2t-1 keys).
    private func _splitChild(parent: BTreeNode<K, V>, childIndex i: Int) {
        let fullChild = parent.children[i]
        let newChild = BTreeNode<K, V>(isLeaf: fullChild.isLeaf)

        let medianKey = fullChild.keys[t - 1]
        let medianValue = fullChild.values[t - 1]

        newChild.keys = Array(fullChild.keys[t...])
        newChild.values = Array(fullChild.values[t...])

        if !fullChild.isLeaf {
            newChild.children = Array(fullChild.children[t...])
            fullChild.children = Array(fullChild.children[..<t])
        }

        fullChild.keys = Array(fullChild.keys[..<(t - 1)])
        fullChild.values = Array(fullChild.values[..<(t - 1)])

        parent.keys.insert(medianKey, at: i)
        parent.values.insert(medianValue, at: i)
        parent.children.insert(newChild, at: i + 1)
    }

    // ── Insert helper ─────────────────────────────────────────────────────────

    @discardableResult
    private func _insertNonFull(node: BTreeNode<K, V>, key: K, value: V) -> Bool {
        let i = node.findKeyIndex(key)
        if i < node.keys.count && node.keys[i] == key {
            node.values[i] = value
            return false
        }
        if node.isLeaf {
            node.keys.insert(key, at: i)
            node.values.insert(value, at: i)
            return true
        }
        var ci = i
        if node.children[ci].keys.count == 2 * t - 1 {
            _splitChild(parent: node, childIndex: ci)
            // After the split, node.keys[ci] is the newly risen median.
            // Decide which of the two halves to descend into.
            if key > node.keys[ci] {
                ci += 1
            } else if key == node.keys[ci] {
                // The key IS the median that just rose into this node.
                node.values[ci] = value
                return false
            }
            // Note: if key < node.keys[ci], ci stays unchanged → go left half.
        }
        return _insertNonFull(node: node.children[ci], key: key, value: value)
    }

    // ── Delete helpers ────────────────────────────────────────────────────────
    //
    // CLRS top-down deletion.  The algorithm has three cases:
    //
    //  Case A: Key k is in node n, and n is a LEAF.  Remove directly.
    //
    //  Case B: Key k is in node n, and n is INTERNAL.
    //    B1: Left child child[i] has ≥ t keys.
    //        Let pred = the in-order predecessor of k (the max key in child[i]).
    //        Replace k with pred in n, then recursively delete pred from child[i].
    //    B2: Right child child[i+1] has ≥ t keys.
    //        Symmetric to B1 using the in-order successor.
    //    B3: Both children have t-1 keys.
    //        Merge child[i], k, and child[i+1] into child[i].
    //        Remove k from child[i] (which now has 2t-1 keys).
    //
    //  Case C: Key k is NOT in node n.  We must descend into some child[i].
    //    Before descending we ensure child[i] has ≥ t keys so that the
    //    recursive deletion cannot underflow it:
    //    C1: A sibling of child[i] has ≥ t keys → rotate through parent.
    //    C2: All siblings have t-1 keys → merge child[i] with a sibling.
    //
    // The helper `_prepareChild(parent:index:)` ensures the child at the given
    // index has ≥ t keys and returns the new index (which may shift left by 1
    // after a left-merge).

    /// Recursively delete `key` from the subtree rooted at `node`.
    ///
    /// INVARIANT (enforced by caller): `node` has ≥ t keys, OR `node` is the
    /// root.  This ensures that if we delete one key from `node` directly
    /// (Case A), node won't underflow.
    ///
    /// Returns true if the key was found anywhere in the subtree.
    private func _delete(node: BTreeNode<K, V>, key: K) -> Bool {
        let i = node.findKeyIndex(key)

        if i < node.keys.count && node.keys[i] == key {
            // ── Key is in THIS node ──────────────────────────────────────────

            if node.isLeaf {
                // Case A: direct removal from leaf.
                // Safe because caller guarantees node has ≥ t keys (or is root).
                node.keys.remove(at: i)
                node.values.remove(at: i)
                return true
            }

            // Internal node.  We need the left and right children of this key.
            let leftChild  = node.children[i]
            let rightChild = node.children[i + 1]

            if leftChild.keys.count >= t {
                // Case B1: Left child is fat enough.
                // Replace key with its in-order predecessor (max of leftChild
                // subtree), then delete that predecessor from leftChild.
                // Pre-fill leftChild's path using the same deletion descent.
                let (predKey, predVal) = _maxKeyValue(node: leftChild)
                node.keys[i]   = predKey
                node.values[i] = predVal
                // Delete predKey from leftChild.  leftChild has ≥ t keys so it
                // satisfies the invariant for the recursive call.
                _ = _deleteDescend(node: leftChild, key: predKey)
                return true

            } else if rightChild.keys.count >= t {
                // Case B2: Right child is fat enough.
                let (succKey, succVal) = _minKeyValue(node: rightChild)
                node.keys[i]   = succKey
                node.values[i] = succVal
                _ = _deleteDescend(node: rightChild, key: succKey)
                return true

            } else {
                // Case B3: Both children have exactly t-1 keys.
                // Merge: pull the key down, absorb rightChild into leftChild.
                _mergeChildren(parent: node, index: i)
                // The merged child is at children[i] and has 2t-1 keys.
                return _delete(node: node.children[i], key: key)
            }

        } else {
            // ── Key is NOT in this node — must descend ────────────────────────
            if node.isLeaf { return false }

            // Before descending into children[i], ensure it has ≥ t keys so
            // that deleting from it won't cause underflow.
            let childIdx = _prepareChild(parent: node, index: i)

            // After _prepareChild a rotation may have moved the key into the
            // parent.  Re-check before descending.
            let j = node.findKeyIndex(key)
            if j < node.keys.count && node.keys[j] == key {
                return _delete(node: node, key: key)
            }
            return _delete(node: node.children[min(childIdx, node.children.count - 1)], key: key)
        }
    }

    /// Delete `key` from the subtree rooted at `node`, pre-filling each node
    /// on the way down to ensure the ≥ t invariant.
    ///
    /// This is exactly `_delete` but used when we know `node` already has ≥ t
    /// keys (Case B1/B2 path).
    private func _deleteDescend(node: BTreeNode<K, V>, key: K) -> Bool {
        return _delete(node: node, key: key)
    }

    /// Ensure that `parent.children[index]` has ≥ t keys.
    /// Returns the new index of the child after possible restructuring.
    ///
    /// If the child already has ≥ t keys, return `index` unchanged.
    /// Otherwise:
    ///   C1a: Borrow from left sibling   → return `index`
    ///   C1b: Borrow from right sibling  → return `index`
    ///   C2a: Merge with left sibling    → return `index - 1` (merged node)
    ///   C2b: Merge with right sibling   → return `index`
    @discardableResult
    private func _prepareChild(parent: BTreeNode<K, V>, index i: Int) -> Int {
        let child = parent.children[i]
        guard child.keys.count < t else { return i }

        let hasLeft  = i > 0
        let hasRight = i < parent.children.count - 1

        if hasLeft && parent.children[i - 1].keys.count >= t {
            // C1a: rotate from left sibling.
            let left = parent.children[i - 1]
            // Push parent separator down to front of child.
            child.keys.insert(parent.keys[i - 1], at: 0)
            child.values.insert(parent.values[i - 1], at: 0)
            // Pull last key from left sibling up to parent.
            parent.keys[i - 1]   = left.keys.removeLast()
            parent.values[i - 1] = left.values.removeLast()
            // Transfer last child pointer if internal.
            if !left.isLeaf {
                child.children.insert(left.children.removeLast(), at: 0)
            }
            return i

        } else if hasRight && parent.children[i + 1].keys.count >= t {
            // C1b: rotate from right sibling.
            let right = parent.children[i + 1]
            child.keys.append(parent.keys[i])
            child.values.append(parent.values[i])
            parent.keys[i]   = right.keys.removeFirst()
            parent.values[i] = right.values.removeFirst()
            if !right.isLeaf {
                child.children.append(right.children.removeFirst())
            }
            return i

        } else if hasLeft {
            // C2a: merge with left sibling.
            // After merge the combined node sits at index i-1.
            _mergeChildren(parent: parent, index: i - 1)
            return i - 1

        } else {
            // C2b: merge with right sibling.
            _mergeChildren(parent: parent, index: i)
            return i
        }
    }

    /// Merge parent.children[i+1] into parent.children[i], pulling down
    /// parent.keys[i] as the separator.
    private func _mergeChildren(parent: BTreeNode<K, V>, index i: Int) {
        let left  = parent.children[i]
        let right = parent.children[i + 1]

        left.keys.append(parent.keys[i])
        left.values.append(parent.values[i])
        left.keys.append(contentsOf: right.keys)
        left.values.append(contentsOf: right.values)
        if !right.isLeaf {
            left.children.append(contentsOf: right.children)
        }
        parent.keys.remove(at: i)
        parent.values.remove(at: i)
        parent.children.remove(at: i + 1)
    }

    private func _maxKeyValue(node: BTreeNode<K, V>) -> (K, V) {
        var n = node
        while !n.isLeaf { n = n.children[n.children.count - 1] }
        return (n.keys[n.keys.count - 1], n.values[n.values.count - 1])
    }

    private func _minKeyValue(node: BTreeNode<K, V>) -> (K, V) {
        var n = node
        while !n.isLeaf { n = n.children[0] }
        return (n.keys[0], n.values[0])
    }

    // ── Traversal helpers ─────────────────────────────────────────────────────

    private func _inorder(node: BTreeNode<K, V>, result: inout [(K, V)]) {
        for i in 0..<node.keys.count {
            if !node.isLeaf { _inorder(node: node.children[i], result: &result) }
            result.append((node.keys[i], node.values[i]))
        }
        if !node.isLeaf { _inorder(node: node.children[node.keys.count], result: &result) }
    }

    private func _rangeQuery(node: BTreeNode<K, V>, low: K, high: K, result: inout [(K, V)]) {
        for i in 0..<node.keys.count {
            if !node.isLeaf && node.keys[i] > low {
                _rangeQuery(node: node.children[i], low: low, high: high, result: &result)
            }
            if node.keys[i] >= low && node.keys[i] <= high {
                result.append((node.keys[i], node.values[i]))
            }
        }
        if !node.isLeaf {
            let last = node.keys.count
            if node.keys[last - 1] < high {
                _rangeQuery(node: node.children[last], low: low, high: high, result: &result)
            }
        }
    }

    // ── Height / validation helpers ───────────────────────────────────────────

    private func _height(node: BTreeNode<K, V>) -> Int {
        if node.isLeaf { return 0 }
        return 1 + _height(node: node.children[0])
    }

    private func _isValid(
        node: BTreeNode<K, V>,
        minKey: K?,
        maxKey: K?,
        expectedDepth: Int,
        currentDepth: Int,
        isRoot: Bool
    ) -> Bool {
        if node.isLeaf && currentDepth != expectedDepth { return false }
        let minKeys = isRoot ? 0 : (t - 1)
        if node.keys.count < minKeys || node.keys.count > 2 * t - 1 { return false }
        if !node.isLeaf && node.children.count != node.keys.count + 1 { return false }
        for i in 0..<node.keys.count {
            if let lo = minKey, node.keys[i] <= lo { return false }
            if let hi = maxKey, node.keys[i] >= hi { return false }
            if i > 0 && node.keys[i] <= node.keys[i - 1] { return false }
        }
        if !node.isLeaf {
            for i in 0...node.keys.count {
                let cMin = i == 0 ? minKey : Optional(node.keys[i - 1])
                let cMax = i == node.keys.count ? maxKey : Optional(node.keys[i])
                if !_isValid(node: node.children[i], minKey: cMin, maxKey: cMax,
                             expectedDepth: expectedDepth, currentDepth: currentDepth + 1,
                             isRoot: false) { return false }
            }
        }
        return true
    }
}

File: code/packages/dart/b-tree/lib/src/b_tree.dart
/// A self-balancing, multi-way search tree (DT11) in pure Dart.
///
/// A B-tree node stores several sorted key-value pairs and, for internal
/// nodes, one more child than key. The [minimumDegree] controls node width:
/// every non-root node has between `t - 1` and `2t - 1` keys. Splitting full
/// children on descent and repairing minimally-filled children before deletion
/// keep every leaf at the same depth.
library b_tree;

/// A sorted key-value entry returned by traversal and range operations.
typedef BTreeEntry<K, V> = (K, V);

Comparator<T> _naturalOrder<T>() => (left, right) =>
    Comparable.compare(left as Comparable, right as Comparable);

final class _BTreeNode<K, V> {
  _BTreeNode({required this.isLeaf});

  final bool isLeaf;
  final List<K> keys = <K>[];
  final List<V> values = <V>[];
  final List<_BTreeNode<K, V>> children = <_BTreeNode<K, V>>[];

  bool isFull(int minimumDegree) => keys.length == 2 * minimumDegree - 1;

  /// Return the first index whose key is greater than or equal to [key].
  int findKeyIndex(K key, Comparator<K> compare) {
    var low = 0;
    var high = keys.length;
    while (low < high) {
      final middle = (low + high) >> 1;
      if (compare(keys[middle], key) < 0) {
        low = middle + 1;
      } else {
        high = middle;
      }
    }
    return low;
  }
}

/// A generic minimum-degree B-tree mapping ordered keys to values.
///
/// Inserting an existing key updates its value without changing [count].
/// [delete] is a no-op that returns `false` when the key is absent. Traversal
/// methods return Dart 3 records in ascending key order.
final class BTree<K, V> {
  /// Create an empty B-tree with minimum degree [minimumDegree].
  ///
  /// The classic 2-3-4 tree uses the default degree of 2. Natural [Comparable]
  /// order is used unless [compare] is supplied.
  BTree([this.minimumDegree = 2, Comparator<K>? compare])
      : _compare = compare ?? _naturalOrder<K>() {
    if (minimumDegree < 2) {
      throw ArgumentError.value(
        minimumDegree,
        'minimumDegree',
        'must be at least 2',
      );
    }
  }

  /// The minimum number of children in a non-root internal node.
  final int minimumDegree;
  final Comparator<K> _compare;

  _BTreeNode<K, V> _root = _BTreeNode<K, V>(isLeaf: true);
  int _count = 0;

  /// Number of key-value pairs stored in the tree.
  int get count => _count;

  /// Whether the tree contains no keys.
  bool get isEmpty => _count == 0;

  /// Number of edges from the root to any leaf.
  int get height {
    var node = _root;
    var result = 0;
    while (!node.isLeaf) {
      node = node.children.first;
      result++;
    }
    return result;
  }

  /// Insert [key] with [value], updating the value when [key] already exists.
  void insert(K key, V value) {
    if (_root.isFull(minimumDegree)) {
      final newRoot = _BTreeNode<K, V>(isLeaf: false)..children.add(_root);
      _splitChild(newRoot, 0);
      _root = newRoot;
    }

    if (_insertNonFull(_root, key, value)) {
      _count++;
    }
  }

  /// Return the value associated with [key], or `null` when it is absent.
  V? search(K key) => _search(_root, key);

  /// Whether [key] is present, including when its stored value is `null`.
  bool contains(K key) => _contains(_root, key);

  /// Remove [key], returning whether it was present.
  ///
  /// This follows the CLRS top-down deletion algorithm. Before descending, a
  /// minimally-filled child borrows from a sibling or merges with one.
  bool delete(K key) {
    // A failed top-down deletion may still merge nodes. Avoid that surprising
    // mutation and the associated root-repair edge case for absent keys.
    if (!contains(key)) return false;

    _deleteRecursive(_root, key);
    _count--;
    if (_root.keys.isEmpty && !_root.isLeaf) {
      _root = _root.children.first;
    }
    return true;
  }

  /// Return the smallest key.
  ///
  /// Throws [StateError] when the tree is empty.
  K minKey() {
    if (isEmpty) throw StateError('Tree is empty.');
    return _minNode(_root).keys.first;
  }

  /// Return the largest key.
  ///
  /// Throws [StateError] when the tree is empty.
  K maxKey() {
    if (isEmpty) throw StateError('Tree is empty.');
    return _maxNode(_root).keys.last;
  }

  /// Return entries whose keys are in the inclusive range [low] through [high].
  List<BTreeEntry<K, V>> rangeQuery(K low, K high) {
    final result = <BTreeEntry<K, V>>[];
    if (_compare(low, high) > 0) return result;
    _collectRange(_root, low, high, result);
    return result;
  }

  /// Return every entry in ascending key order.
  List<BTreeEntry<K, V>> inorder() {
    final result = <BTreeEntry<K, V>>[];
    _collectInorder(_root, result);
    return result;
  }

  /// Verify key-count, ordering, child-count, value, and leaf-depth invariants.
  bool isValid() {
    if (_count == 0) {
      return _root.isLeaf &&
          _root.keys.isEmpty &&
          _root.values.isEmpty &&
          _root.children.isEmpty;
    }
    if (inorder().length != _count) return false;
    final leafDepth = <int?>[null];
    return _validate(
      _root,
      minKey: null,
      hasMinKey: false,
      maxKey: null,
      hasMaxKey: false,
      depth: 0,
      leafDepth: leafDepth,
      isRoot: true,
    );
  }

  V? _search(_BTreeNode<K, V> node, K key) {
    final index = node.findKeyIndex(key, _compare);
    if (index < node.keys.length && _compare(node.keys[index], key) == 0) {
      return node.values[index];
    }
    return node.isLeaf ? null : _search(node.children[index], key);
  }

  bool _contains(_BTreeNode<K, V> node, K key) {
    final index = node.findKeyIndex(key, _compare);
    if (index < node.keys.length && _compare(node.keys[index], key) == 0) {
      return true;
    }
    return !node.isLeaf && _contains(node.children[index], key);
  }

  void _splitChild(_BTreeNode<K, V> parent, int childIndex) {
    final child = parent.children[childIndex];
    final right = _BTreeNode<K, V>(isLeaf: child.isLeaf);
    final middle = minimumDegree - 1;
    final medianKey = child.keys[middle];
    final medianValue = child.values[middle];

    right.keys.addAll(child.keys.sublist(middle + 1));
    right.values.addAll(child.values.sublist(middle + 1));
    if (!child.isLeaf) {
      right.children.addAll(child.children.sublist(minimumDegree));
      child.children.removeRange(minimumDegree, child.children.length);
    }

    child.keys.removeRange(middle, child.keys.length);
    child.values.removeRange(middle, child.values.length);
    parent.keys.insert(childIndex, medianKey);
    parent.values.insert(childIndex, medianValue);
    parent.children.insert(childIndex + 1, right);
  }

  bool _insertNonFull(_BTreeNode<K, V> node, K key, V value) {
    var index = node.findKeyIndex(key, _compare);
    if (index < node.keys.length && _compare(node.keys[index], key) == 0) {
      node.values[index] = value;
      return false;
    }

    if (node.isLeaf) {
      node.keys.insert(index, key);
      node.values.insert(index, value);
      return true;
    }

    if (node.children[index].isFull(minimumDegree)) {
      _splitChild(node, index);
      final comparison = _compare(key, node.keys[index]);
      if (comparison == 0) {
        node.values[index] = value;
        return false;
      }
      if (comparison > 0) index++;
    }
    return _insertNonFull(node.children[index], key, value);
  }

  void _deleteRecursive(_BTreeNode<K, V> node, K key) {
    final index = node.findKeyIndex(key, _compare);
    final found =
        index < node.keys.length && _compare(node.keys[index], key) == 0;

    if (found) {
      if (node.isLeaf) {
        node.keys.removeAt(index);
        node.values.removeAt(index);
        return;
      }

      final left = node.children[index];
      final right = node.children[index + 1];
      if (left.keys.length >= minimumDegree) {
        final predecessor = _maxNode(left);
        final predecessorKey = predecessor.keys.last;
        node.keys[index] = predecessorKey;
        node.values[index] = predecessor.values.last;
        _deleteRecursive(left, predecessorKey);
      } else if (right.keys.length >= minimumDegree) {
        final successor = _minNode(right);
        final successorKey = successor.keys.first;
        node.keys[index] = successorKey;
        node.values[index] = successor.values.first;
        _deleteRecursive(right, successorKey);
      } else {
        final merged = _mergeChildren(node, index);
        _deleteRecursive(merged, key);
      }
      return;
    }

    if (node.isLeaf) return;
    final childIndex = _ensureMinKeys(node, index);
    _deleteRecursive(node.children[childIndex], key);
  }

  _BTreeNode<K, V> _mergeChildren(_BTreeNode<K, V> parent, int leftIndex) {
    final left = parent.children[leftIndex];
    final right = parent.children[leftIndex + 1];

    left.keys.add(parent.keys.removeAt(leftIndex));
    left.values.add(parent.values.removeAt(leftIndex));
    parent.children.removeAt(leftIndex + 1);
    left.keys.addAll(right.keys);
    left.values.addAll(right.values);
    if (!left.isLeaf) left.children.addAll(right.children);
    return left;
  }

  int _ensureMinKeys(_BTreeNode<K, V> parent, int childIndex) {
    final child = parent.children[childIndex];
    if (child.keys.length >= minimumDegree) return childIndex;

    if (childIndex > 0) {
      final left = parent.children[childIndex - 1];
      if (left.keys.length >= minimumDegree) {
        child.keys.insert(0, parent.keys[childIndex - 1]);
        child.values.insert(0, parent.values[childIndex - 1]);
        parent.keys[childIndex - 1] = left.keys.removeLast();
        parent.values[childIndex - 1] = left.values.removeLast();
        if (!left.isLeaf) child.children.insert(0, left.children.removeLast());
        return childIndex;
      }
    }

    if (childIndex < parent.children.length - 1) {
      final right = parent.children[childIndex + 1];
      if (right.keys.length >= minimumDegree) {
        child.keys.add(parent.keys[childIndex]);
        child.values.add(parent.values[childIndex]);
        parent.keys[childIndex] = right.keys.removeAt(0);
        parent.values[childIndex] = right.values.removeAt(0);
        if (!right.isLeaf) child.children.add(right.children.removeAt(0));
        return childIndex;
      }
    }

    if (childIndex > 0) {
      _mergeChildren(parent, childIndex - 1);
      return childIndex - 1;
    }
    _mergeChildren(parent, childIndex);
    return childIndex;
  }

  _BTreeNode<K, V> _minNode(_BTreeNode<K, V> node) {
    while (!node.isLeaf) {
      node = node.children.first;
    }
    return node;
  }

  _BTreeNode<K, V> _maxNode(_BTreeNode<K, V> node) {
    while (!node.isLeaf) {
      node = node.children.last;
    }
    return node;
  }

  void _collectInorder(_BTreeNode<K, V> node, List<BTreeEntry<K, V>> result) {
    if (node.isLeaf) {
      for (var index = 0; index < node.keys.length; index++) {
        result.add((node.keys[index], node.values[index]));
      }
      return;
    }

    for (var index = 0; index < node.keys.length; index++) {
      _collectInorder(node.children[index], result);
      result.add((node.keys[index], node.values[index]));
    }
    _collectInorder(node.children.last, result);
  }

  void _collectRange(
    _BTreeNode<K, V> node,
    K low,
    K high,
    List<BTreeEntry<K, V>> result,
  ) {
    var index = 0;
    while (index < node.keys.length) {
      final key = node.keys[index];
      if (!node.isLeaf && _compare(low, key) < 0) {
        _collectRange(node.children[index], low, high, result);
      }
      if (_compare(key, high) > 0) return;
      if (_compare(key, low) >= 0) {
        result.add((key, node.values[index]));
      }
      index++;
    }
    if (!node.isLeaf) {
      _collectRange(node.children[index], low, high, result);
    }
  }

  bool _validate(
    _BTreeNode<K, V> node, {
    required K? minKey,
    required bool hasMinKey,
    required K? maxKey,
    required bool hasMaxKey,
    required int depth,
    required List<int?> leafDepth,
    required bool isRoot,
  }) {
    final keyCount = node.keys.length;
    if (node.values.length != keyCount) return false;
    if (keyCount > 2 * minimumDegree - 1) return false;
    if (isRoot) {
      if (_count > 0 && keyCount < 1) return false;
    } else if (keyCount < minimumDegree - 1) {
      return false;
    }

    for (var index = 0; index < keyCount; index++) {
      final key = node.keys[index];
      if (hasMinKey && _compare(key, minKey as K) <= 0) return false;
      if (hasMaxKey && _compare(key, maxKey as K) >= 0) return false;
      if (index > 0 && _compare(key, node.keys[index - 1]) <= 0) {
        return false;
      }
    }

    if (node.isLeaf) {
      if (node.children.isNotEmpty) return false;
      leafDepth[0] ??= depth;
      return leafDepth[0] == depth;
    }
    if (node.children.length != keyCount + 1) return false;

    for (var index = 0; index <= keyCount; index++) {
      final childMin = index > 0 ? node.keys[index - 1] : minKey;
      final childMax = index < keyCount ? node.keys[index] : maxKey;
      if (!_validate(
        node.children[index],
        minKey: childMin,
        hasMinKey: index > 0 || hasMinKey,
        maxKey: childMax,
        hasMaxKey: index < keyCount || hasMaxKey,
        depth: depth + 1,
        leafDepth: leafDepth,
        isRoot: false,
      )) {
        return false;
      }
    }
    return true;
  }

  @override
  String toString() => 'BTree(t=$minimumDegree, size=$count, height=$height)';
}
